src/lib.rs

```diff
@@ -285,6 +288,25 @@ impl WalkDir {
         self.opts.max_open = n;
         self
     }
+
+    /// Set a compare function.
+    /// If a compare function is set, WalkDir will return all paths in sorted
+    /// order.
+    ///
+    /// ```rust,no-run
+    /// use std::cmp;
+    /// use std::ffi::OsString;
+    /// use walkdir::WalkDir;
+    ///
+    /// fn string_compare(a: &OsString, b: &OsString) -> cmp::Ordering {
+    ///     a.cmp(b)
+    /// }
+    /// WalkDir::new("foo").sort(Some(string_compare));
+    /// ```
+    pub fn sort(mut self, sorter: Option<fn(&OsString,&OsString) -> std::cmp::Ordering>) -> Self {
```

Comment:
So I'm not quite sure what you mean with avoiding calling the same function twice, but I think this should work:

```rust
pub fn sort_by<F: FnMut(&OsString, &OsString) -> Ordering>(mut self, cmp: F) -> Self {
    self.opts.sorter = Some(Box::new(cmp));
    self
}
```

And you'll need this in the `WalkDirOptions` type:

```rust
struct WalkDirOptions {
    // ...
    sorter: Option<Box<FnMut(&OsString,&OsString) -> std::cmp::Ordering>>,
```

I think that should do it!